File: templates/python/06-rag-knowledge-server/src/lib/logger.py
"""Structured logger that writes to stderr.

CRITICAL: MCP servers using stdio transport must NEVER write to stdout.
stdout is reserved for JSON-RPC messages. Any log output to stdout will
corrupt the protocol stream and break the client connection.

This logger writes JSON lines to stderr, which clients capture and display
in their debug/developer consoles.
"""

from __future__ import annotations

import json
import sys
from datetime import UTC, datetime
from typing import Any


LogLevel = str  # "debug" | "info" | "warn" | "error"


def _log(level: LogLevel, message: str, meta: dict[str, Any] | None = None) -> None:
    entry: dict[str, Any] = {
        "level": level,
        "message": message,
        "timestamp": datetime.now(UTC).isoformat(),
    }
    if meta:
        entry.update(meta)
    sys.stderr.write(json.dumps(entry, default=str) + "\n")
    sys.stderr.flush()


class Logger:
    def debug(self, msg: str, meta: dict[str, Any] | None = None) -> None:
        _log("debug", msg, meta)

    def info(self, msg: str, meta: dict[str, Any] | None = None) -> None:
        _log("info", msg, meta)

    def warn(self, msg: str, meta: dict[str, Any] | None = None) -> None:
        _log("warn", msg, meta)

    def error(self, msg: str, meta: dict[str, Any] | None = None) -> None:
        _log("error", msg, meta)


logger = Logger()

File: templates/typescript/05-web-search-fetch/src/lib/logger.ts
/**
 * Structured logger that writes to stderr.
 *
 * CRITICAL: MCP servers using stdio transport must NEVER write to stdout.
 * stdout is reserved for JSON-RPC messages. Any log output to stdout
 * will corrupt the protocol stream and break the client connection.
 *
 * This logger writes JSON to stderr, which clients capture and display
 * in their debug/developer consoles.
 */

type LogLevel = "debug" | "info" | "warn" | "error";

function log(level: LogLevel, message: string, meta?: Record<string, unknown>) {
  const entry = {
    level,
    message,
    timestamp: new Date().toISOString(),
    ...meta,
  };
  process.stderr.write(JSON.stringify(entry) + "\n");
}

export const logger = {
  debug: (msg: string, meta?: Record<string, unknown>) => log("debug", msg, meta),
  info: (msg: string, meta?: Record<string, unknown>) => log("info", msg, meta),
  warn: (msg: string, meta?: Record<string, unknown>) => log("warn", msg, meta),
  error: (msg: string, meta?: Record<string, unknown>) => log("error", msg, meta),
};
